Implement initial support for Rule-Id violations

StarKhan6368 opened the pull request:
**What**: Implement initial support for Rule-Id violations

**Why**: Each violation is now assigned a Rule-Id (e.g., R0001) to enable users to quickly reference its precise definition and suggested resolution in the rules documentation, and allowing the UI to deliver better reporting

**Checklist**:

- [x] Unit Tests
- [x] Build passing locally
- [ ] Sonar Quality Gate
- [ ] Security scans don't report any vulnerabilities
- [ ] Documentation added/updated (share link)
- [ ] Sample Project added/updated (share link)
- [ ] Demo video (share link)
- [ ] Article on Website (share link)
- [ ] Roadmpap updated (share link)
- [ ] Conference Talk (share link)

StarKhan6368 committed:
Implement rule violation details

- Each Result.Failure can have RuleViolationId
- RuleViolationId stores context and violation breadCrumb
- Can then be finalized to a rule-id linking to documentation

StarKhan6368 committed:
Add StandardRuleViolationSegment for key checks

- Add Context for Test and Stub

StarKhan6368 committed:
Ensure all scalars are calling the correct mismatchResult

- Ensure parse and matches call misMatch result leading to
  Parse and ValueMisMatch RuleViolationSegment respectively

StarKhan6368 committed:
Simplify RuleViolation Implementation

- Remove Context and add multi-violation support

StarKhan6368 committed:
Add OpenApi Specific RuleViolations

StarKhan6368 committed:
Configure rest of the patterns to call correct failure helpers

StarKhan6368 committed:
Implement Results to Errors Conversion

StarKhan6368 committed:
Provide rawResult instead of details to testHooks

- Add RuleViolation Snapshots to help with serialization
- Ensure RuleViolations are not duplicated in the report

StarKhan6368 committed:
Provide rawResult instead of details to mockHooks

StarKhan6368 committed:
Add comprehensive tests for BreadCrumbToJsonPathConverter

joelrosario committed:
Fixed duplicate id and downcased the id in the rule url

joelrosario committed:
Minor changes to the rule ids

StarKhan6368 committed:
Fix all the failing tests

- Update reportString assertions to take RuleId into account

StarKhan6368 committed:
Minor refactorings and renaming

StarKhan6368 committed:
Add tests for RuleViolations

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/Discriminator.kt:
```diff
@@ -164,32 +170,32 @@ class Discriminator(
             "Discriminator property $discriminatorProperty is missing from the object (it's value should be $discriminatorCsv)",
             breadCrumb = discriminatorProperty,
-            failureReason = FailureReason.DiscriminatorMismatch
+            failureReason = FailureReason.DiscriminatorMismatch,
```
`failureReason` and `ruleViolation` appear very similar if you just look at the names. However, they serve different purposes, and maybe the names could make this clearer.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/QueryParameterScalarPattern.kt:
```diff
@@ -17,23 +17,17 @@ data class QueryParameterScalarPattern(override val pattern: Pattern): Pattern b
-            Result.Failure(exceptionCauseMessage(e))
-        }
+        val parsedValue = runCatching { pattern.parse(sampleDataString, resolver) }.getOrDefault(StringValue(sampleDataString))
+        return resolver.matchesPattern(null, pattern, parsedValue)
```
is it possible for matchesPattern to throw an exception? If it does, it will bubble up as an exception, rather than a failure, since the try-catch has been removed.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/AnyPattern.kt:
```diff
@@ -214,7 +214,8 @@ data class AnyPattern(
         val failuresWithUpdatedBreadcrumbs = addTypeInfoBreadCrumbs(matchResults)
 
-        return Result.fromFailures(failuresWithUpdatedBreadcrumbs)
+        val didNotMatchAnyOptionFailure = Failure(message = "Value didn't match any of the schema options", ruleViolation = StandardRuleViolation.ONE_OF_VALUE_MISMATCH)
```
We can add the rule violation directly to the composite result.

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/AnyPattern.kt:
```diff
@@ -339,7 +340,7 @@ data class AnyPattern(
 
         return if(compatibleResult is Failure && allValuesAreScalar())
-            mismatchResult(this, otherPattern, thisResolver.mismatchMessages)
+            patternMismatchResult(this, otherPattern, thisResolver.mismatchMessages)
```
Use datatype mismatch helper

joelrosario commented on core/src/main/kotlin/io/specmatic/core/pattern/DateTimePattern.kt:
```diff
@@ -50,7 +49,7 @@ object DateTimePattern : Pattern, ScalarType {
         get() = null
 
-    override val typeName: String = "datetime"
+    override val typeName: String = "date-time"
```
This should go back to "datetime"

StarKhan6368 committed:
Resolve Review feedback

- Revert modifications to typeName for patterns
- Dont emit seperate error for ONE_OF_VALUE_MISMATCH
- Use TYPE_MISMATCH for format mismatches when matching
- Update tests for the same

StarKhan6368 committed:
Revert to the older error message for scalar queryParam

- When actual is an array instead of single value
- TitleCase summary in ruleViolation Text

joelrosario commented on core/src/test/kotlin/io/specmatic/core/HttpStubTests.kt:
```diff
@@ -843,18 +876,50 @@ Scenario: JSON API to get account details with fact check
-            Expected number, actual was "ABC"
-            >> RESPONSE.BODY.petId
-            Expected number, actual was "DEF"
+            ${
```
The errors that show up seem confusing. Seems like 3 errors, but it's actually the same mismatch causing two errors..?

joelrosario commented on core/src/test/kotlin/io/specmatic/core/TestBackwardCompatibilityKtTest.kt:
```diff
@@ -1253,7 +1253,7 @@ paths:
 
         assertThat(result.report()).contains("string in the new contract")
-        assertThat(result.report()).contains("number in the old contract")
+        assertThat(result.report()).contains("(number) in the old contract")
```
how come string has no brackets but "(number)" does?

StarKhan6368 committed:
Remove Path and SecurityScheme OpenApiRuleViolations

- Add appropriate RuleViolations OpenApiSecuritySchemes matches methods
- Same with MultiPartFormData matching logic
- Update tests assertions accordingly

StarKhan6368 committed:
Revert encompass failure message in EmptyStringPattern

- Remove inaccurate TODO as the behavior is actually expected

StarKhan6368 committed:
Revamp MismatchMesssages for consistent error messages

- Use `Specification` instead of `contract` or `stub` keywords
- Remove unecessary mismatch messages, consolidate the variations

StarKhan6368 committed:
Update test expectations and setups for mismatchMessages

- Instead of hardcoding the messaage utilize the expecte mismachMessage
  implementation by invoking the correct messagae function

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Rename schema key labels to property instead of key

- Update test setups / assertions for the same

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix MismatchMessage formatting for keyLabels

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix BackwardCompat check messages

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix AnyPattern parse logic to rethrow errors

- Remove ONE_OF_VALUE_MISMATCH and reword summaries

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Add VALUE_MISMATCH to ContentType mismatch

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Disable RuleViolations of BCC and Strict Stub

- Remove ruleViolations from the final result

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Add special mismatchMessages for StubExample

- Fix tests and update tests for the same

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix JsonArrayPattern to use mismatchMessages

- Use mismatch message helper for length violation

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix StrictMode400 Response to be more rich in result

- Use NamedMismatch messages for examples if possible
- Rich Result.Failure instead of rawString

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Add Composite schema test for RuleViolations

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Add toNativeValue method to Value interface

- Converts Value to native variants like map, list, string, etc

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

StarKhan6368 committed:
Fix Edge case in BreadCrumbToJsonPathConverter

- Rename variable in attemptParse for better clarity

Signed-off-by: Sufiyan <StarKhan6368@gmail.com>

joelrosario committed:
Updated version to 2.35.0 as it will be a feature release

joelrosario committed:
Merge pull request #2145 from specmatic/consistent-messages

Revise MismatchMessages for consistent error messages

joelrosario committed:
Merge branch 'main' into rule-id-test-and-stub

joelrosario committed:
Fixed failing test that was missed earlier
